### ISSUE TITLE
Indentation

### ISSUE THREAD

**Issue by guibou** [guibou/PyF#13]
Current PyF implementation does nothing related to initial whitespaces in string. For example:

```
>>> putStrLn [f|Account informations:
  - Owner: {owner}
  - Balance : {balance:.2}|]
Account informations:
  - Owner: Guillaume
  - Balance : 3.14
```

is not satisfying on many points:

- First line appears directly after the `[f|` symbol and is not really visible
- First line is not indented as other lines
- Others lines are less indented than the surrounding code.
- End of quasiquote `|]` must appears at the end of the last line

I want to be able to write something like that:

```
>>> putStrLn [f|
    Account informations:
      - Owner: {owner}
      - Balance : {balance:.2}
  |]
```

In this solution, the block of text is clearly separated from the quasi quote symbols and the indentation inside the block is coherent. It can be also indented with respect the surrounding code, for example in case of deeply nested code, compare:

```
f a b = case a of
   _ pat -> if cond
       then [f|- a
- b|]
       else ""
   otherPat -> ...
```

with

```
f a b = case a of
   _ pat -> if cond
       then [f|
           - a
           - b
       |]
       else ""
   otherPat -> ...
```

This follows the python docstring convention, detailed here: https://www.python.org/dev/peps/pep-0257/#handling-docstring-indentation which can be resumed as:

- First and last lines are ignored if they only contains whitespaces
- Indentation is based on the minimum amount of whitespaces

I'm tempted to implement that, but many questions are still opened:

- Should this be a separated quasi quoter (`fTrim` for trimmed, or `f` for trimmed and `fRaw` for untrimmed) to explicit the different behaviors?
- Users won't be able to start / end a multiline string by empty lines. This does not appears to be a huge limitation.
- There is no way to write a formatting string in which each lines are indented by a fixed amount. Is that important?

**Comment by ghorn** [comment]
Why follow the docstring convention instead of the multiline string convention? For me it would be clearest and most flexible if
`ghci >>> putStrLn [f|xxx|]`
was exactly equivalent to
`python3 >>> print("""xxx""")`
for any contents of `xxx` including newlines and `\`

You could get the first-line-indented-with-the-rest behavior with a `\`, for example:
```
ghci >>> putStrLn [f|\
  a
  b
  c
|]
```
In my use-case I would prefer to have exact control of leading indentation and not have to make an additional pass over the `Text`, for instance
```
ghci >>> putStrLn $ unlines $ map ("  " <>) $ lines $ [f|\
a
b
c
|]
```
seems less readable, and less efficient as well.

What is the rational for striping leading indentation?

**Comment by guibou** [comment]
@ghorn the rational for stripping leading indentation was to have beautiful source file. I'll implement that later as a standalone function, for example;

```
foo = trim [f|\
      - a
      - b
      - c\
      |]
```

Will remove indentation based on the minimum indentation of the block.

Escaping of `\` was implemented in 0.7.1.0. I'm closing the issue.
